appsec-kit-backend/src/main/java/com/vaadin/appsec/backend/AppSecService.java

```diff
@@ -43,28 +61,152 @@ public static AppSecService getInstance() {
         return AppSecService.InstanceHolder.instance;
     }
 
+    private final List<AppSecScanEventListener> scanEventListeners = new ArrayList<>();
+
+    private final OpenSourceVulnerabilityService osvService;
+
+    private final VulnerabilityStore vulnerabilityStore;
+
+    private final BillOfMaterialsStore bomStore;
+
+    private final AppSecDTOProvider dtoProvider;
+
     private AppSecConfiguration configuration;
 
     private AppSecData data;
 
+    private Clock clock = Clock.systemUTC();
+
+    private ScheduledFuture<?> scheduledScan;
+
     /**
      * Creates a new instance of the service.
      *
      * @param configuration
      *            the configuration bean
      */
     private AppSecService(AppSecConfiguration configuration) {
+        bomStore = new BillOfMaterialsStore();
+        osvService = new OpenSourceVulnerabilityService();
+        vulnerabilityStore = new VulnerabilityStore(osvService, bomStore);
+        dtoProvider = new AppSecDTOProvider(vulnerabilityStore, bomStore);
         this.configuration = configuration;
     }
 
+    /**
+     * Initializes the service reading the SBOM file.
+     */
+    public void init() {
+        cancelScheduledScan();
+        Path bomFilePath = configuration.getBomFilePath();
+        try {
+            bomStore.readBomFile(bomFilePath);
+        } catch (ParseException e) {
+            throw new AppSecException(
+                    "Cannot parse the SBOM file: " + bomFilePath.toString(), e);
+        }
+        readOrCreateDataFile();
+    }
+
+    /**
+     * Schedules automatic scan for vulnerabilities at a fixed rate set to the
+     * value configured with
+     * {@link AppSecConfiguration#setAutoScanInterval(java.time.Duration)}.
+     */
+    public void scheduleAutomaticScan() {
+        checkForInitialization();
+        long initialDelay = 0;
+        long autoScanPeriod = configuration.getAutoScanInterval().getSeconds();
+        Instant lastScan = data.getLastScan();
+        if (lastScan != null) {
+            long secondsUntilNextScan = autoScanPeriod
+                    - lastScan.until(clock.instant(), ChronoUnit.SECONDS);
+            if (secondsUntilNextScan > 0) {
+                initialDelay = secondsUntilNextScan;
+            }
+        }
+        scheduledScan = configuration.getTaskExecutor()
+                .scheduleAtFixedRate(() -> {
+                    vulnerabilityStore.refresh();
```

Comment:
At this point the `refresh()` can throw a `NullPointerException` because the `bom` inside the function can be still not initialized. This can happen if the public `AppSecService.getInstance().getData()` is called before the `scheduleAutomaticScan()` or the `scanForVulnerabilities()` public APIs. The `getData()` function creates the `data` and the `checkForInitialization()` will run successfully in the above public APIs while the `AppSecService.getInstance().init()` which creates the `bom` is never called. Could this be a problem?

Reply:
The `checkForInitialization()` should also check if the `init()` function has been run.

Reply:
I don't follow, how the BOM can be still no initialized if we require initialization before scheduling the job?

Reply:
We require only that the `AppSecData` is not null in the `checkForInitialization()` but we never check if the `BillOfMaterialsStore.bom` is initialized through the `AppSecService.init()`.
In theory if the
`AppSecService.getInstance().getData()`
`AppSecService.getInstance().scheduleAutomaticScan()`
or the
`AppSecService.getInstance().getData()`
`AppSecService.getInstance().scanForVulnerabilities()`
public APIs are called after each other and never calling the `AppSecService.init()`, a NPE can happen.

Reply:
That's true, now also the BOM is required not to be null. Thanks!

Reply:
I see there were changes in the [5531ba2](https://github.com/vaadin/appsec-kit/pull/38/commits/5531ba2f1c05a1079c383bb2bdfdab2e823c3951) commit regarding the `AppSecService.init()` call. Now that the `init()` is called before the scans in the `AppSecServiceInitListener.serviceInit()` my above assumption can't happen anymore because that guarantees that the `init()` is always called when the appsec-kit starts. This is for v8 version, if we can achieve the same in v7 version then we are good.
However, I think we can keep the check what you added that the BOM is not null.